### PR TITLE
feat: allow to overwrite java release version, closes MISC-367

### PR DESCRIPTION
**Pull request by tpetillot**
**Motivation:**
Some project doesn't require java 1.8

**Modifications:**
Add setting for java release version. When not default, doesn't set
related options (configuration is left to user)

### CODE REVIEW

File: src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala [hunk 1]
```diff
@@ -25,16 +25,28 @@ object GatlingCompilerSettingsPlugin extends AutoPlugin {
...
+  import autoImport._
+
   private val isJava8 = scala.util.Properties.javaVersion.startsWith("1.8")
 
-  private val JavacOptions =
+  private val JavacOptions = Def.setting {
+    val version = javaReleaseVersion.value
     if (isJava8) {
-      Seq("-source", "1.8", "-target", "1.8")
+      Seq("-source", version, "-target", version)
```
Comment (slandelle) [inline src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala:42]: Why???

File: src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala [hunk 2]
```diff
@@ -43,26 +55,26 @@ object GatlingCompilerSettingsPlugin extends AutoPlugin {
...
         "-unchecked",
         "-language:implicitConversions",
-        "-target:jvm-1.8"
       )
 
     if (isJava8) {
       base
     } else {
-      base ++ Seq("-release", "8")
+      base ++ Seq("-release", version)
```
Comment (slandelle) [inline src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala:63]: Wrong

https://docs.scala-lang.org/overviews/compiler-options/index.html

> Compile for a specific version of the Java platform. Supported targets: 6, 7, 8, 9

**Review by slandelle — CHANGES_REQUESTED** [review]

File: src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala [hunk 3]
```diff
@@ -25,28 +25,46 @@ object GatlingCompilerSettingsPlugin extends AutoPlugin {
...
-        "-target:jvm-1.8"
+        "-language:implicitConversions"
       )
 
-    if (isJava8) {
+    // When user overwrite java release version, we leave him the responsibility
+    if (isJava8 || releaseVersion != DefaultJavaReleaseVersion) {
       base
     } else {
       base ++ Seq("-release", "8")
```
Comment (slandelle) [inline src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala:70]: What's the difference between `-target` and `-release`???

> -release:
> Compile for a specific version of the Java platform. Supported targets: 6, 7, 8, 9

> -target:
> Target platform for object files. ([8],9,10,11,12)

**Review by slandelle — CHANGES_REQUESTED** [review]

File: src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala (hunk 3 above)
Comment (tpetillot) [reply src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala:70]: A good explanation: https://stackoverflow.com/a/43103038
> javac, --release N is equivalent to -source N -target N -bootclasspath <bootclasspath-from-N>

**Review by slandelle — APPROVED** [review]

File: src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala [hunk 4]
```diff
@@ -25,44 +25,68 @@ object GatlingCompilerSettingsPlugin extends AutoPlugin {
...
+  private val JavacOptions = Def.setting {
+    val version = javaReleaseVersion.value
+    if (version == DefaultJavaReleaseVersion) {
+      if (isJava8) {
+        Seq("-source", "1.8", "-target", "1.8")
+      } else {
+        Seq("--release", "1.8")
+      }
     } else {
-      Seq("--release", "8")
```
Comment (Isammoc) [inline src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala:34]: Are you sure that we want to move from `8` to `1.8`?
I do think it is the other way (`target` and `release` should accept `8`, `9`...`12`)

File: src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala [hunk 5]
```diff
@@ -25,44 +25,68 @@ object GatlingCompilerSettingsPlugin extends AutoPlugin {
...
+      // When user overwrite java release version, we leave him the responsibility
```
Comment (Isammoc) [inline src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala:49]: As javacOptions are defined only with that. I won't bother to have a special case for that.
Plugin users (module) can already override such configuration.

File: src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala [hunk 6]
```diff
@@ -25,44 +25,68 @@ object GatlingCompilerSettingsPlugin extends AutoPlugin {
...
       base
     } else {
-      base ++ Seq("-release", "8")
+      val baseWithTarget = base :+ "-target:jvm-1.8"
+      if (isJava8) {
+        baseWithTarget
+      } else {
+        baseWithTarget ++ Seq("-release", "8")
+      }
     }
```
Comment (Isammoc) [inline src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala:76]: I personally, would have create a

```scala
gatlingScalacJavaVersion := if (isJava8) {
  Seq("-target:jvm-1.8")
} else {
  Seq("-target:jvm-1.8", "-target", "8")
}
```

And adding this settings to the `scalacOptions`
That allows plugin users to only override that.

Comment (slandelle) [reply src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala:76]: ?! Scalac is very likely to either choke or only honor the last value. Is this something you've tested?

File: src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala (hunk 4 above)
Comment (slandelle) [reply src/main/scala/io/gatling/build/compile/GatlingCompilerSettingsPlugin.scala:34]: > I do think it is the other way (target and release should accept 8, 9...12)

To be tested. Maybe 8 only works with Scala 3?

**Comment by tpetillot** [comment]
Closed, not a priority for now. Related issue have been sent to backlog.

**Pull request closed without merging**